// @astryxdesign/theme-neutral/dist/chunk-TC7YR7RU.mjs
// src/icons.tsx
import {
  X,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  Check,
  CheckCircle,
  XCircle,
  AlertTriangle,
  Info,
  Calendar,
  Clock,
  ExternalLink,
  Menu,
  MoreHorizontal,
  Search,
  ArrowUp,
  ArrowDown,
  ArrowUpDown,
  Filter,
  EyeOff,
  Columns,
  Copy,
  CheckCheck,
  Wrench,
  Square,
  Mic
} from "lucide-react";
import { jsx } from "react/jsx-runtime";
var iconProps = {
  size: "1em",
  "aria-hidden": true
};
var neutralIconRegistry = {
  close: /* @__PURE__ */ jsx(X, { ...iconProps }),
  chevronDown: /* @__PURE__ */ jsx(ChevronDown, { ...iconProps }),
  chevronLeft: /* @__PURE__ */ jsx(ChevronLeft, { ...iconProps }),
  chevronRight: /* @__PURE__ */ jsx(ChevronRight, { ...iconProps }),
  check: /* @__PURE__ */ jsx(Check, { ...iconProps }),
  success: /* @__PURE__ */ jsx(CheckCircle, { ...iconProps }),
  error: /* @__PURE__ */ jsx(XCircle, { ...iconProps }),
  warning: /* @__PURE__ */ jsx(AlertTriangle, { ...iconProps }),
  info: /* @__PURE__ */ jsx(Info, { ...iconProps }),
  calendar: /* @__PURE__ */ jsx(Calendar, { ...iconProps }),
  clock: /* @__PURE__ */ jsx(Clock, { ...iconProps }),
  externalLink: /* @__PURE__ */ jsx(ExternalLink, { ...iconProps }),
  menu: /* @__PURE__ */ jsx(Menu, { ...iconProps }),
  moreHorizontal: /* @__PURE__ */ jsx(MoreHorizontal, { ...iconProps }),
  search: /* @__PURE__ */ jsx(Search, { ...iconProps }),
  arrowUp: /* @__PURE__ */ jsx(ArrowUp, { ...iconProps }),
  arrowDown: /* @__PURE__ */ jsx(ArrowDown, { ...iconProps }),
  arrowsUpDown: /* @__PURE__ */ jsx(ArrowUpDown, { ...iconProps }),
  funnel: /* @__PURE__ */ jsx(Filter, { ...iconProps }),
  eyeSlash: /* @__PURE__ */ jsx(EyeOff, { ...iconProps }),
  viewColumns: /* @__PURE__ */ jsx(Columns, { ...iconProps }),
  copy: /* @__PURE__ */ jsx(Copy, { ...iconProps }),
  checkDouble: /* @__PURE__ */ jsx(CheckCheck, { ...iconProps }),
  wrench: /* @__PURE__ */ jsx(Wrench, { ...iconProps }),
  stop: /* @__PURE__ */ jsx(Square, { ...iconProps }),
  microphone: /* @__PURE__ */ jsx(Mic, { ...iconProps })
};

export {
  neutralIconRegistry
};
