// @astryxdesign/core/dist/Badge/Badge.test-violations.js
// Copyright (c) Meta Platforms, Inc. and affiliates.

/**
 * @file Test file with intentional violations for ESLint plugin testing
 * Run with: pnpm lint packages/core/src/Badge/Badge.test-violations.tsx
 */

import * as stylex from '@stylexjs/stylex';

// This file contains intentional violations to test the Astryx ESLint plugin

const badStyles = {
  hardcodedFontSize: {
    kGuDYH: "xif65rj",
    $$css: true
  },
  hardcodedFontWeight: {
    k63SB2: "x1s688f",
    $$css: true
  },
  hardcodedColor: {
    kMwMTN: "xj8ax7g",
    kWkggS: "xcpsgoo",
    $$css: true
  },
  hardcodedSpacing: {
    kmVPX3: "x1tamke2",
    kogj98: "xkib98w",
    kOIVth: "x1jnr06f",
    $$css: true
  },
  hardcodedRadius: {
    kaIpWk: "xur7f20",
    $$css: true
  },
  missingLetterSpacing: {
    kGuDYH: "xfifm61",
    k63SB2: "xk50ysn",
    $$css: true
  },
  goodStyles: {
    k1xSpc: "x78zum5",
    kGNEyG: "x6s0dn4",
    kjj79g: "xl56j7k",
    $$css: true
  },
  allowedValues: {
    kmVPX3: "x1717udv",
    kogj98: "x1bpp3o7",
    kMwMTN: "x1heor9g",
    $$css: true
  }
};
export { badStyles };
