// @astryxdesign/core/dist/CodeBlock/CodeBlock.js
// Copyright (c) Meta Platforms, Inc. and affiliates.

'use client';

/**
 * @file CodeBlock.tsx
 * @input Uses React, StyleX, theme tokens, CSS Custom Highlight API
 * @output Exports CodeBlock component and CodeBlockProps
 * @position Core implementation; read-only syntax-highlighted code display
 */
import { useInsertionEffect, useEffect, useRef, useState, useCallback, useMemo } from 'react';
import * as React from 'react';
import * as stylex from '@stylexjs/stylex';
import "../theme/tokens.stylex.js";
import { colorVars, spacingVars, radiusVars, typographyVars, fontWeightVars, typeScaleVars, borderVars, durationVars, easeVars } from "../theme/tokens.stylex.js";
import { mergeProps } from "../utils/index.js";
import { Icon } from "../Icon/index.js";
import { tokenize, tokenizeAsync, flatTokensToLines, SYNC_TOKENIZE_THRESHOLD } from "./tokenizer.js";
import { ensureHighlightStyles } from "./highlightStyles.js";
import { applyHighlightRangesChunked } from "./highlightRanges.js";
import { themeProps } from "../utils/themeProps.js";

// ---------------------------------------------------------------------------
// Styles
// ---------------------------------------------------------------------------
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
const containerStyles = {
  card: {
    kaIpWk: "xh6dtrn",
    kMzoRj: "x1litavf",
    ksu8eU: "x1y0btm7",
    kVAM5u: "x14i3s5s",
    $$css: true
  },
  section: {
    kaIpWk: "x2u8bby",
    kMzoRj: "xc342km",
    ksu8eU: "xng3xce",
    kVAM5u: "x9r1u3d",
    "--color-syntax-background": "x1mtvhk1",
    $$css: true
  }
};
const dynamicStyles = {
  width: value => [{
    kzqmXN: value != null ? "x5lhr3w" : value,
    k7Eaqz: (value === 'fit-content' ? 'min(100%, 400px)' : null) != null ? "xkj4a21" : value === 'fit-content' ? 'min(100%, 400px)' : null,
    ks0D6T: (value === 'fit-content' ? '100%' : null) != null ? "xf68679" : value === 'fit-content' ? '100%' : null,
    $$css: true
  }, {
    "--x-width": (val => typeof val === "number" ? val + "px" : val != null ? val : undefined)(value),
    "--x-minWidth": (val => typeof val === "number" ? val + "px" : val != null ? val : undefined)(value === 'fit-content' ? 'min(100%, 400px)' : null),
    "--x-maxWidth": (val => typeof val === "number" ? val + "px" : val != null ? val : undefined)(value === 'fit-content' ? '100%' : null)
  }]
};
const styles = {
  root: {
    kVAEAm: "x1n2onr6",
    kHBbk8: "xc8icb0",
    k1xSpc: "x78zum5",
    kXwgrk: "xdt5ytf",
    kogj98: "x1ghz6dp",
    kWkggS: "xzyc27v",
    kVQacm: "xb3r6kr",
    $$css: true
  },
  gutter: {
    kmuXW: "x2lah0s",
    k8WAf4: "x8o8v82",
    kZCmMZ: "x1rey3nv",
    kwRFfy: "x1t818jl",
    k9WMMc: "xp4054r",
    kfSwDN: "x87ps6o",
    kMwMTN: "xt5rlkq",
    kWqL5O: "x1pcaw5z",
    kSWEuD: "x32b0ac",
    k26BEO: "xtgwc6q",
    $$css: true
  },
  code: {
    k1xSpc: "x1lliihq",
    kUk6DE: "x98rzlu",
    k8WAf4: "x8o8v82",
    kg3NbH: "x1pzlopt",
    kogj98: "x1ghz6dp",
    kMv6JI: "x9m5x89",
    kMwMTN: "xdfisg",
    kKi2Bq: "xeq4nuv",
    khDVqt: "x1sdyfia",
    kTgw9: "x1lldw8n",
    kHjlTd: "x1h4wwuj",
    $$css: true
  },
  codeWrapped: {
    khDVqt: "x126k92a",
    kTgw9: "x1yn0g08",
    kHjlTd: "x1mzt3pk",
    $$css: true
  },
  sizeSm: {
    kGuDYH: "x141an7d",
    $$css: true
  },
  sizeMd: {
    kGuDYH: "xp03k98",
    $$css: true
  },
  gutterSm: {
    kGuDYH: "x141an7d",
    $$css: true
  },
  gutterMd: {
    kGuDYH: "xp03k98",
    $$css: true
  }
};

// ---------------------------------------------------------------------------
// Line rendering
// ---------------------------------------------------------------------------

const LINE_CHUNK_SIZE = 20;
const LINE_CHUNK_THRESHOLD = 100;

/**
 * Memoized chunk component — cheaper than memoizing every individual line.
 */
const CodeChunk = /*#__PURE__*/React.memo(function CodeChunk({
  lines,
  startIndex,
  highlightSet,
  renderLineContent
}) {
  return /*#__PURE__*/_jsx(_Fragment, {
    children: lines.map((line, j) => {
      const i = startIndex + j;
      return /*#__PURE__*/_jsx("div", {
        "data-line": i + 1,
        ...{
          0: {
            className: "x17iicif"
          },
          1: {
            className: "x17iicif xgcxg3y x1ng9b9t x1pzlopt"
          }
        }[!!(highlightSet?.has(i + 1) ?? false) << 0],
        children: renderLineContent(line, i)
      }, i);
    })
  });
});
function renderLines(lines, highlightSet, renderLineContent, chunkSize = LINE_CHUNK_SIZE) {
  chunkSize = Math.max(1, Math.floor(chunkSize));
  if (lines.length < LINE_CHUNK_THRESHOLD) {
    return /*#__PURE__*/_jsx(CodeChunk, {
      lines: lines,
      startIndex: 0,
      highlightSet: highlightSet,
      renderLineContent: renderLineContent
    });
  }
  const chunks = [];
  for (let start = 0; start < lines.length; start += chunkSize) {
    const end = Math.min(start + chunkSize, lines.length);
    const chunkLines = lines.slice(start, end);
    const estimatedHeight = `${chunkLines.length}lh`;
    chunks.push(/*#__PURE__*/_jsx("div", {
      ...mergeProps({
        className: "xb5mbof"
      }, {
        style: {
          containIntrinsicBlockSize: `auto ${estimatedHeight}`
        }
      }),
      children: /*#__PURE__*/_jsx(CodeChunk, {
        lines: chunkLines,
        startIndex: start,
        highlightSet: highlightSet,
        renderLineContent: renderLineContent
      })
    }, start));
  }
  return chunks;
}

// ---------------------------------------------------------------------------
// Props
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function hasHighlightAPI() {
  return typeof CSS !== 'undefined' && 'highlights' in CSS && typeof Highlight !== 'undefined';
}

/**
 * Safari supports the Highlight API JS objects but has rendering issues
 * with ::highlight() in code blocks. Detect Safari (WebKit without Chrome)
 * so we can fall back to spans.
 */
function isSafari() {
  if (typeof navigator === 'undefined') {
    return false;
  }
  const ua = navigator.userAgent;
  return ua.includes('AppleWebKit') && !ua.includes('Chrome');
}

/**
 * Hook: per-line tokens with sync/async + custom tokenizer compat.
 */
function useTokenLines(code, language, customTokenizer) {
  const [asyncTokenResult, setAsyncTokenResult] = useState(null);
  const syncTokens = useMemo(() => {
    if (customTokenizer) {
      return flatTokensToLines(customTokenizer(code, language), code);
    }
    if (code.length >= SYNC_TOKENIZE_THRESHOLD) {
      return null;
    }
    return tokenize(code, language);
  }, [code, language, customTokenizer]);
  useEffect(() => {
    if (code.length < SYNC_TOKENIZE_THRESHOLD || customTokenizer) {
      return;
    }
    const abortController = new AbortController();
    async function tokenizeLargeCode() {
      try {
        const tokens = await tokenizeAsync(code, language, abortController.signal);
        if (!abortController.signal.aborted) {
          setAsyncTokenResult({
            code,
            language,
            tokens
          });
        }
      } catch {
        if (!abortController.signal.aborted) {
          setAsyncTokenResult({
            code,
            language,
            tokens: []
          });
        }
      }
    }
    void tokenizeLargeCode();
    return () => {
      abortController.abort();
    };
  }, [code, language, customTokenizer]);
  if (syncTokens != null) {
    return syncTokens;
  }
  if (asyncTokenResult?.code === code && asyncTokenResult.language === language) {
    return asyncTokenResult.tokens;
  }
  return [];
}

// ---------------------------------------------------------------------------
// Span-mode code element
// ---------------------------------------------------------------------------

function buildSpanLine(lineText, tokens) {
  if (tokens.length === 0) {
    return lineText || '\u200b';
  }
  const parts = [];
  let cursor = 0;
  for (const token of tokens) {
    if (token.start > cursor) {
      parts.push(lineText.slice(cursor, token.start));
    }
    const end = Math.min(token.end, lineText.length);
    parts.push(/*#__PURE__*/_jsx("span", {
      className: `astryx-token-${token.type} xds-token-${token.type}`,
      children: lineText.slice(token.start, end)
    }, `${token.start}-${token.type}`));
    cursor = end;
  }
  if (cursor < lineText.length) {
    parts.push(lineText.slice(cursor));
  }
  return parts.length > 0 ? parts : '\u200b';
}
function SpanCodeContent({
  lines,
  tokenLines,
  highlightSet,
  isWrapped,
  sizeStyle
}) {
  useInsertionEffect(() => {
    ensureHighlightStyles();
  }, []);
  const renderLineContent = useCallback((line, lineIndex) => {
    const tokens = tokenLines[lineIndex] ?? [];
    return buildSpanLine(line, tokens);
  }, [tokenLines]);
  return /*#__PURE__*/_jsx("code", {
    ...stylex.props(styles.code, sizeStyle, isWrapped && styles.codeWrapped),
    children: renderLines(lines, highlightSet, renderLineContent)
  });
}

// ---------------------------------------------------------------------------
// Range-mode code element
// ---------------------------------------------------------------------------

function RangeCodeContent({
  lines,
  tokenLines,
  highlightSet,
  isWrapped,
  sizeStyle
}) {
  const codeRef = useRef(null);
  useEffect(() => {
    if (!hasHighlightAPI()) {
      return;
    }
    ensureHighlightStyles();
    const codeEl = codeRef.current;
    if (!codeEl || tokenLines.length === 0) {
      return;
    }
    return applyHighlightRangesChunked(codeEl, tokenLines);
  }, [tokenLines]);
  const renderLineContent = useCallback(line => line || '\u200b', []);
  return /*#__PURE__*/_jsx("code", {
    ref: codeRef,
    ...stylex.props(styles.code, sizeStyle, isWrapped && styles.codeWrapped),
    children: renderLines(lines, highlightSet, renderLineContent)
  });
}

// ---------------------------------------------------------------------------
// Main component
// ---------------------------------------------------------------------------

/**
 * Read-only code display with syntax highlighting, line numbers,
 * and optional copy button.
 *
 * @example
 * ```
 * <CodeBlock code="const x = 42;" language="javascript" />
 * ```
 */
export function CodeBlock({
  code,
  language = 'plaintext',
  title,
  hasLanguageLabel = true,
  hasLineNumbers = false,
  highlightLines,
  hasCopyButton = true,
  onCopy,
  isWrapped = false,
  maxHeight,
  isCollapsible = false,
  collapsibleThreshold = 10,
  size = 'md',
  width: widthProp = 'fit-content',
  container = 'card',
  tokenizer: customTokenizer,
  highlightMode = 'auto',
  xstyle,
  className,
  style,
  ref,
  ...props
}) {
  const [copied, setCopied] = useState(false);
  const useSpans = highlightMode === 'spans' || highlightMode === 'auto' && !hasHighlightAPI() || highlightMode === 'auto' && isSafari();
  const lines = useMemo(() => {
    const l = code.split('\n');
    if (l.length > 1 && l[l.length - 1] === '') {
      l.pop();
    }
    return l;
  }, [code]);
  const tokenLines = useTokenLines(code, language, customTokenizer);
  const highlightSet = useMemo(() => highlightLines ? new Set(highlightLines) : null, [highlightLines]);
  const handleCopy = useCallback(async () => {
    try {
      await navigator.clipboard.writeText(code);
      setCopied(true);
      onCopy?.();
      setTimeout(() => setCopied(false), 2000);
    } catch {
      // Clipboard failures leave the copied state unchanged.
    }
  }, [code, onCopy]);
  const sizeStyle = size === 'sm' ? styles.sizeSm : styles.sizeMd;
  const gutterSizeStyle = size === 'sm' ? styles.gutterSm : styles.gutterMd;
  const languageLabel = hasLanguageLabel && language !== 'plaintext' ? language : null;
  const showHeader = title != null || languageLabel != null;
  const canCollapse = isCollapsible && lines.length >= collapsibleThreshold;
  const [isCollapsed, setIsCollapsed] = useState(false);
  const scrollContainerRef = useRef(null);
  const scrollStyle = maxHeight ? {
    maxHeight: typeof maxHeight === 'number' ? `${maxHeight}px` : maxHeight
  } : undefined;
  const copyIcon = /*#__PURE__*/_jsx(Icon, {
    icon: copied ? 'check' : 'copy',
    size: "sm",
    color: "inherit"
  });
  const copyButtonEl = hasCopyButton ? /*#__PURE__*/_jsx("button", {
    type: "button",
    onClick: e => {
      // Stop propagation so copying does not toggle the collapsible header.
      e.stopPropagation();
      void handleCopy();
    },
    "aria-label": copied ? 'Copied' : 'Copy code',
    ...{
      0: {
        className: "x3nfvp2 x6s0dn4 xl56j7k x9epnlk x3kzqx6 xx3sua9 xjbqb8w xe9uy6x x1mjzq1q x1ypdohk x14ju556"
      },
      1: {
        className: "x3nfvp2 x6s0dn4 xl56j7k x9epnlk x3kzqx6 xx3sua9 xjbqb8w xe9uy6x x1mjzq1q x1ypdohk x14ju556 x10l6tqk xctzyg x1n93rvh"
      }
    }[!!!showHeader << 0],
    children: copyIcon
  }) : null;
  const headerEl = showHeader ? /*#__PURE__*/_jsxs("div", {
    ...{
      0: {
        className: "x78zum5 x6s0dn4 x1qughib x1pzlopt xzyc27v x7wzq59 x13vifvy x1vjfegm xce4md1"
      },
      1: {
        className: "x78zum5 x6s0dn4 x1qughib x1pzlopt xzyc27v x7wzq59 x13vifvy x1vjfegm xce4md1 x92x3c3 x1q0q8m5 xw8gpjh"
      }
    }[!!hasLineNumbers << 0],
    children: [/*#__PURE__*/_jsx("div", {
      role: canCollapse ? 'button' : undefined,
      tabIndex: canCollapse ? 0 : undefined,
      "aria-expanded": canCollapse ? !isCollapsed : undefined,
      onClick: canCollapse ? () => setIsCollapsed(prev => !prev) : undefined,
      onKeyDown: canCollapse ? e => {
        if (e.key === 'Enter' || e.key === ' ') {
          e.preventDefault();
          setIsCollapsed(prev => !prev);
        }
      } : undefined,
      ...{
        0: {
          className: "x78zum5 x6s0dn4 x98rzlu xeuugli x1717udv xjbqb8w x1heor9g xln7xf2 x1yc453h"
        },
        1: {
          className: "x78zum5 x6s0dn4 x98rzlu xeuugli x1717udv xjbqb8w x1heor9g xln7xf2 x1yc453h x1ypdohk x87ps6o"
        }
      }[!!canCollapse << 0],
      children: /*#__PURE__*/_jsxs("span", {
        ...{
          className: "x78zum5 x6s0dn4 xzye2dw x141an7d x9m5x89 x1e4wzip x1mjzq1q x1ghz6dp x1ltkj2j"
        },
        children: [title, title && languageLabel ? ' — ' : '', languageLabel, canCollapse && /*#__PURE__*/_jsx("span", {
          ...{
            0: {
              className: "x3nfvp2 x6s0dn4 xl56j7k x2lah0s x6jxa94 x1v9usgg x1mjzq1q x11xpdln xuedmi6 xlr8y92"
            },
            1: {
              className: "x3nfvp2 x6s0dn4 xl56j7k x2lah0s x6jxa94 x1v9usgg x1mjzq1q x11xpdln xuedmi6 xlr8y92 x19jd1h0"
            }
          }[!!isCollapsed << 0],
          children: /*#__PURE__*/_jsx(Icon, {
            icon: "chevronDown",
            size: "xsm",
            color: "inherit"
          })
        })]
      })
    }), copyButtonEl]
  }) : null;
  const codeBody = /*#__PURE__*/_jsx("div", {
    ref: scrollContainerRef
    // The scroll container is keyboard-focusable so keyboard users can
    // scroll long or wide code that overflows the viewport. Uses
    // role="group" (not "region") so multiple code blocks on a page don't
    // create duplicate same-named landmarks (axe: landmark-unique).
    ,
    tabIndex: 0,
    role: "group",
    "aria-label": languageLabel ?? 'Code',
    ...mergeProps({
      className: "xw2csxc x1odjw0f"
    }, {
      style: scrollStyle
    }),
    children: /*#__PURE__*/_jsxs("div", {
      ...{
        0: {
          className: "x78zum5 x1pshirs"
        },
        1: {
          className: "x78zum5 x1pshirs x1ajqlyv"
        }
      }[!!(showHeader && !hasLineNumbers) << 0],
      children: [hasLineNumbers && /*#__PURE__*/_jsx("div", {
        ...stylex.props(styles.gutter, gutterSizeStyle),
        "aria-hidden": "true",
        children: lines.map((_, i) =>
        /*#__PURE__*/
        // eslint-disable-next-line @eslint-react/no-array-index-key -- gutter line numbers are positional by definition
        _jsx("div", {
          ...{
            className: "x9m5x89 x17iicif"
          },
          children: i + 1
        }, i))
      }), useSpans ? /*#__PURE__*/_jsx(SpanCodeContent, {
        lines: lines,
        tokenLines: tokenLines,
        highlightSet: highlightSet,
        isWrapped: isWrapped,
        sizeStyle: sizeStyle
      }) : /*#__PURE__*/_jsx(RangeCodeContent, {
        lines: lines,
        tokenLines: tokenLines,
        highlightSet: highlightSet,
        isWrapped: isWrapped,
        sizeStyle: sizeStyle
      })]
    })
  });
  return /*#__PURE__*/_jsxs("pre", {
    ref: ref,
    ...mergeProps(themeProps('codeblock', {
      size,
      language,
      container
    }), stylex.props(styles.root, dynamicStyles.width(widthProp), containerStyles[container], xstyle), className, style),
    ...props,
    children: [headerEl, canCollapse ? /*#__PURE__*/_jsx("div", {
      ...{
        0: {
          className: "xrvj5dj x1tu4anv x1qn9uv2 x80gvsz xlr8y92"
        },
        1: {
          className: "xrvj5dj x1qn9uv2 x80gvsz xlr8y92 xihq33y"
        }
      }[!!isCollapsed << 0],
      children: /*#__PURE__*/_jsx("div", {
        ...{
          className: "xb3r6kr x2lwn1j"
        },
        children: codeBody
      })
    }) : codeBody, !showHeader && copyButtonEl]
  });
}
CodeBlock.displayName = 'CodeBlock';
